render-test/manifest_parser.cpp

```diff
@@ -0,0 +1,392 @@
+                               "Invalid expectationPath is provoided inside the manifest file: %s",
+                               filePath.c_str());
+            return {};
+        }
+        for (const auto& value : expectationPathValue.GetArray()) {
+            if (!value.IsString()) {
+                return {};
+            }
+            expectationPaths.emplace_back(getValidPath(manifest.manifestPath, value.GetString()));
+            if (expectationPaths.back().empty()) {
+                return {};
+            }
+        }
+    }
+    std::vector<mbgl::filesystem::path> ignorePaths{};
+    if (document.HasMember("ignore_paths")) {
+        const auto& ignorePathValue = document["ignore_paths"];
+        if (!ignorePathValue.IsArray()) {
+            mbgl::Log::Warning(
+                mbgl::Event::General, "Invalid ignorePath is provoided inside the manifest file: %s", filePath.c_str());
+            return {};
+        }
+        for (const auto& value : ignorePathValue.GetArray()) {
+            if (!value.IsString()) {
+                return {};
+            }
+            ignorePaths.emplace_back(getValidPath(manifest.manifestPath, value.GetString()));
+            if (ignorePaths.back().empty()) {
+                return {};
+            }
+        }
+        manifest.ignores = parseIgnores(ignorePaths);
+    }
+
+    bool enbaleProbeTest{false};
+    std::vector<mbgl::filesystem::path> paths;
+    // Seperate Probe tests and normal tests, since they have different testRoot
+    if (std::find(testNames.begin(), testNames.end(), "tests") != testNames.end()) {
+        paths.emplace_back(probeTestPath / "tests");
+        enbaleProbeTest = true;
+    } else {
+        for (const auto& id : testNames) {
+            paths.emplace_back(baseTestPath / id);
+        }
+        if (paths.empty()) {
+            paths.emplace_back(baseTestPath);
+        }
+    }
+
+    // Recursively traverse through the test paths and collect test dirπectories containing "style.son".
+    auto& testPaths = manifest.testPaths;
+    testPaths.reserve(paths.size());
+    for (const auto& path : paths) {
+        if (!mbgl::filesystem::exists(path)) {
+            mbgl::Log::Warning(
+                mbgl::Event::General, "Provided test folder '%s' does not exist.", path.string().c_str());
+            continue;
+        }
+        for (auto& testPath : mbgl::filesystem::recursive_directory_iterator(path)) {
+            // Skip paths that fail regexp match.
+            if (!testFilter.empty() && !std::regex_match(testPath.path().string(), std::regex(testFilter))) {
+                continue;
+            }
+
+            if (testPath.path().filename() == "style.json") {
+                testPaths.emplace_back(testPath, getTestExpectations(testPath, path, expectationPaths));
+            }
+        }
+    }
+
+    manifest.testRootPath = enbaleProbeTest ? probeTestPath.string() : baseTestPath.string();
+    if (manifest.testRootPath.back() == '/') {
```

Comment:
is this condition really needed? I'd assume `mbgl::filesystem` will return string in the same way all the time :thinking:

Reply:
Nope, it will return the original path string, which may contain "/" or not.